// @gomomento/sdk-core@1.106.0/dist/src/utils/collection-ttl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollectionTtl = void 0;
const utils_1 = require("../internal/utils");
const time_1 = require("./time");
/** Represents the desired behavior for managing the TTL on collection
 *  objects (dictionaries, lists, sets) in your cache.
 *
 *  For cache operations that modify a collection, there are a few things
 *  to consider.  The first time the collection is created, we need to
 *  set a TTL on it.  For subsequent operations that modify the collection
 *  you may choose to update the TTL in order to prolong the life of the
 *  cached collection object, or you may choose to leave the TTL unmodified
 *  in order to ensure that the collection expires at the original TTL.
 *
 *  The default behavior is to refresh the TTL (to prolong the life of the
 *  collection) each time it is written.  This behavior can be modified
 *  by calling CollectionTtl.withNoRefreshTtlOnUpdates().
 *
 *  A null TTL means to use the client's TTL.
 */
class CollectionTtl {
    /**
     * If refreshTtl is true, the client must update the collection's TTL
     * when it modifies a collection.
     * A null ttl means to use the client's TTL.
     * @param {number | null} [ttlSeconds=null]
     * @param {boolean} [refreshTtl=true]
     */
    constructor(ttlSeconds = null, refreshTtl = true) {
        if (ttlSeconds !== null) {
            (0, utils_1.validateTtlSeconds)(ttlSeconds);
        }
        this._refreshTtl = refreshTtl;
        this._ttlSeconds = ttlSeconds;
    }
    /** Time-to-live, in seconds.
     * @returns {number | null}
     */
    ttlSeconds() {
        return this._ttlSeconds;
    }
    /** Time-to-live, in milliseconds.
     * @returns {number | null}
     */
    ttlMilliseconds() {
        return this._ttlSeconds === null
            ? null
            : (0, time_1.secondsToMilliseconds)(this._ttlSeconds);
    }
    /** Whether or not to refresh a collection's TTL when it's modified.
     * @returns {boolean}
     */
    refreshTtl() {
        return this._refreshTtl;
    }
    /** The default way to handle TTLs for collections. The client's default TTL
     *  will be used, and the TTL for the collection will be refreshed any
     *  time the collection is modified.
     * @constructor
     * @returns {CollectionTtl}
     */
    static fromCacheTtl() {
        return new CollectionTtl(null, true);
    }
    /** Constructs a CollectionTtl with the specified TTL. The TTL
     *  for the collection will be refreshed any time the collection is
     *  modified.
     * @constructor
     * @param {number} [ttlSeconds]
     * @returns {CollectionTtl}
     */
    static of(ttlSeconds) {
        return new CollectionTtl(ttlSeconds, true);
    }
    /** Constructs a CollectionTtl with the specified TTL.
     *  Will only refresh if the TTL is provided.
     * @constructor
     * @param {number | null} [ttlSeconds=null]
     * @returns {CollectionTtl}
     */
    static refreshTtlIfProvided(ttlSeconds = null) {
        return new CollectionTtl(ttlSeconds, ttlSeconds !== null);
    }
    /** Copies the CollectionTtl, but it will refresh the TTL when
     *  the collection is modified.
     * @returns {CollectionTtl}
     */
    withRefreshTtlOnUpdates() {
        return new CollectionTtl(this._ttlSeconds, true);
    }
    /** Copies the CollectionTTL, but the TTL will not be refreshed
     *  when the collection is modified. Use this if you want to ensure
     *  that your collection expires at the originally specified time, even
     *  if you make modifications to the value of the collection.
     * @returns {CollectionTtl}
     */
    withNoRefreshTtlOnUpdates() {
        return new CollectionTtl(this._ttlSeconds, false);
    }
    /** A string represenation of the CollectionTtl for debugging purposes.
     * @return {CollectionTtl}
     */
    toString() {
        return `ttl: ${this._ttlSeconds || 'null'}, refreshTtl: ${this._refreshTtl ? 'true' : 'false'}`;
    }
}
exports.CollectionTtl = CollectionTtl;
