// @gomomento/sdk-core@1.106.0/dist/src/internal/clients/storage/AbstractStorageClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractStorageClient = void 0;
class AbstractStorageClient {
    constructor(dataClients, controlClient) {
        this.dataClients = dataClients;
        this.controlClient = controlClient;
        // We round-robin the requests through all of our clients.  Since javascript
        // is single-threaded, we don't have to worry about thread safety on this
        // index variable.
        this.nextDataClientIndex = 0;
    }
    createStore(storeName) {
        return this.controlClient.createStore(storeName);
    }
    listStores() {
        return this.controlClient.listStores();
    }
    deleteStore(storeName) {
        return this.controlClient.deleteStore(storeName);
    }
    get(storeName, key) {
        return this.getNextDataClient().get(storeName, key);
    }
    putInt(storeName, key, value) {
        return this.getNextDataClient().putInt(storeName, key, value);
    }
    putDouble(storeName, key, value) {
        return this.getNextDataClient().putDouble(storeName, key, value);
    }
    putString(storeName, key, value) {
        return this.getNextDataClient().putString(storeName, key, value);
    }
    putBytes(storeName, key, value) {
        return this.getNextDataClient().putBytes(storeName, key, value);
    }
    delete(storeName, key) {
        return this.getNextDataClient().delete(storeName, key);
    }
    getNextDataClient() {
        const client = this.dataClients[this.nextDataClientIndex];
        this.nextDataClientIndex =
            (this.nextDataClientIndex + 1) % this.dataClients.length;
        return client;
    }
}
exports.AbstractStorageClient = AbstractStorageClient;
