// @gomomento/sdk@1.106.0/dist/src/config/middleware/request-logging-middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestLoggingMiddleware = void 0;
const __1 = require("../..");
class LoggingMiddlewareRequestHandler {
    constructor(logger, requestId) {
        this.logger = logger;
        this.requestId = requestId;
    }
    onRequestMetadata(metadata) {
        this.logger.info(JSON.stringify({
            event: 'onRequestMetadata',
            requestId: this.requestId,
            ...metadata.toJsonObject(),
        }));
        return Promise.resolve(metadata);
    }
    onRequestBody(request) {
        this.logger.info(JSON.stringify({
            event: 'onRequestBody',
            requestId: this.requestId,
            requestSize: request.messageLength(),
            ...request.toLogFormat(),
        }));
        return Promise.resolve(request);
    }
    onResponseMetadata(metadata) {
        this.logger.info(JSON.stringify({
            event: 'onResponseMetadata',
            requestId: this.requestId,
            ...metadata.toJsonObject(),
        }));
        return Promise.resolve(metadata);
    }
    onResponseBody(response) {
        var _a;
        this.logger.info(JSON.stringify({
            event: 'onResponseBody',
            requestId: this.requestId,
            ...(_a = response === null || response === void 0 ? void 0 : response._grpcMessage) === null || _a === void 0 ? void 0 : _a.toObject(),
        }));
        return Promise.resolve(response);
    }
    onResponseStatus(status) {
        this.logger.info(JSON.stringify({
            event: 'onResponseStatus',
            requestId: this.requestId,
            status: status.code(),
        }));
        return Promise.resolve(status);
    }
}
/**
 * This middleware implements per-request logging which can be used for
 * debugging.
 *
 * WARNING: enabling this middleware may have minor performance implications,
 * so enable with caution.
 *
 * In order for this middleware to produce output you will need to have
 * set up your {Configuration} with a {MomentoLoggerFactory} instance that
 * is configured to log at DEBUG level or lower.  See `advanced.ts` in the
 * examples directory for an example of how to set up your {Configuration} to
 * enable this middleware.
 */
class RequestLoggingMiddleware {
    constructor(loggerFactory = new __1.DefaultMomentoLoggerFactory()) {
        this.logger = loggerFactory.getLogger(this);
        this.nextRequestId = 0;
    }
    onNewRequest() {
        this.nextRequestId++;
        return new LoggingMiddlewareRequestHandler(this.logger, this.nextRequestId.toString());
    }
}
exports.RequestLoggingMiddleware = RequestLoggingMiddleware;
