// @gomomento/sdk-core@1.106.0/dist/src/internal/utils/semaphore.js
"use strict";
// Source: https://gist.github.com/gregkorossy/e33be1f201cf242197d9c4d0a1fa7335
Object.defineProperty(exports, "__esModule", { value: true });
exports.Semaphore = void 0;
class Semaphore {
    constructor(max) {
        this.counter = 0;
        this.waiting = [];
        this.max = max;
    }
    take() {
        if (this.waiting.length > 0 && this.counter < this.max) {
            this.counter += 1;
            const promise = this.waiting.shift();
            if (promise) {
                promise.resolve();
            }
        }
    }
    acquire() {
        if (this.counter < this.max) {
            this.counter += 1;
            return new Promise(resolve => {
                resolve(undefined);
            });
        }
        return new Promise((resolve, err) => {
            this.waiting.push({ resolve, err });
        });
    }
    release() {
        this.counter -= 1;
        this.take();
    }
    purge() {
        const unresolved = this.waiting.length;
        for (let i = 0; i < unresolved; i += 1) {
            this.waiting[i].err('Task has been purged.');
        }
        this.counter = 0;
        this.waiting = [];
        return unresolved;
    }
}
exports.Semaphore = Semaphore;
