// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/enums/leaderboard/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeaderboardUpsertResponse = exports.LeaderboardRemoveElementsResponse = exports.LeaderboardLengthResponse = exports.LeaderboardFetchResponse = exports.LeaderboardDeleteResponse = void 0;
var LeaderboardDeleteResponse;
(function (LeaderboardDeleteResponse) {
    LeaderboardDeleteResponse["Success"] = "Success";
    LeaderboardDeleteResponse["Error"] = "Error";
})(LeaderboardDeleteResponse = exports.LeaderboardDeleteResponse || (exports.LeaderboardDeleteResponse = {}));
var LeaderboardFetchResponse;
(function (LeaderboardFetchResponse) {
    LeaderboardFetchResponse["Success"] = "Success";
    LeaderboardFetchResponse["Error"] = "Error";
})(LeaderboardFetchResponse = exports.LeaderboardFetchResponse || (exports.LeaderboardFetchResponse = {}));
var LeaderboardLengthResponse;
(function (LeaderboardLengthResponse) {
    LeaderboardLengthResponse["Success"] = "Success";
    LeaderboardLengthResponse["Error"] = "Error";
})(LeaderboardLengthResponse = exports.LeaderboardLengthResponse || (exports.LeaderboardLengthResponse = {}));
var LeaderboardRemoveElementsResponse;
(function (LeaderboardRemoveElementsResponse) {
    LeaderboardRemoveElementsResponse["Success"] = "Success";
    LeaderboardRemoveElementsResponse["Error"] = "Error";
})(LeaderboardRemoveElementsResponse = exports.LeaderboardRemoveElementsResponse || (exports.LeaderboardRemoveElementsResponse = {}));
var LeaderboardUpsertResponse;
(function (LeaderboardUpsertResponse) {
    LeaderboardUpsertResponse["Success"] = "Success";
    LeaderboardUpsertResponse["Error"] = "Error";
})(LeaderboardUpsertResponse = exports.LeaderboardUpsertResponse || (exports.LeaderboardUpsertResponse = {}));
