// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/enums/cache/set/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheSetLengthResponse = exports.CacheSetPopResponse = exports.CacheSetSampleResponse = exports.CacheSetRemoveElementsResponse = exports.CacheSetRemoveElementResponse = exports.CacheSetContainsElementsResponse = exports.CacheSetContainsElementResponse = exports.CacheSetFetchResponse = exports.CacheSetAddElementsResponse = exports.CacheSetAddElementResponse = void 0;
var CacheSetAddElementResponse;
(function (CacheSetAddElementResponse) {
    CacheSetAddElementResponse["Success"] = "Success";
    CacheSetAddElementResponse["Error"] = "Error";
})(CacheSetAddElementResponse = exports.CacheSetAddElementResponse || (exports.CacheSetAddElementResponse = {}));
var CacheSetAddElementsResponse;
(function (CacheSetAddElementsResponse) {
    CacheSetAddElementsResponse["Success"] = "Success";
    CacheSetAddElementsResponse["Error"] = "Error";
})(CacheSetAddElementsResponse = exports.CacheSetAddElementsResponse || (exports.CacheSetAddElementsResponse = {}));
var CacheSetFetchResponse;
(function (CacheSetFetchResponse) {
    CacheSetFetchResponse["Hit"] = "Hit";
    CacheSetFetchResponse["Miss"] = "Miss";
    CacheSetFetchResponse["Error"] = "Error";
})(CacheSetFetchResponse = exports.CacheSetFetchResponse || (exports.CacheSetFetchResponse = {}));
var CacheSetContainsElementResponse;
(function (CacheSetContainsElementResponse) {
    CacheSetContainsElementResponse["Hit"] = "Hit";
    CacheSetContainsElementResponse["Miss"] = "Miss";
    CacheSetContainsElementResponse["Error"] = "Error";
})(CacheSetContainsElementResponse = exports.CacheSetContainsElementResponse || (exports.CacheSetContainsElementResponse = {}));
var CacheSetContainsElementsResponse;
(function (CacheSetContainsElementsResponse) {
    CacheSetContainsElementsResponse["Hit"] = "Hit";
    CacheSetContainsElementsResponse["Miss"] = "Miss";
    CacheSetContainsElementsResponse["Error"] = "Error";
})(CacheSetContainsElementsResponse = exports.CacheSetContainsElementsResponse || (exports.CacheSetContainsElementsResponse = {}));
var CacheSetRemoveElementResponse;
(function (CacheSetRemoveElementResponse) {
    CacheSetRemoveElementResponse["Success"] = "Success";
    CacheSetRemoveElementResponse["Error"] = "Error";
})(CacheSetRemoveElementResponse = exports.CacheSetRemoveElementResponse || (exports.CacheSetRemoveElementResponse = {}));
var CacheSetRemoveElementsResponse;
(function (CacheSetRemoveElementsResponse) {
    CacheSetRemoveElementsResponse["Success"] = "Success";
    CacheSetRemoveElementsResponse["Error"] = "Error";
})(CacheSetRemoveElementsResponse = exports.CacheSetRemoveElementsResponse || (exports.CacheSetRemoveElementsResponse = {}));
var CacheSetSampleResponse;
(function (CacheSetSampleResponse) {
    CacheSetSampleResponse["Hit"] = "Hit";
    CacheSetSampleResponse["Miss"] = "Miss";
    CacheSetSampleResponse["Error"] = "Error";
})(CacheSetSampleResponse = exports.CacheSetSampleResponse || (exports.CacheSetSampleResponse = {}));
var CacheSetPopResponse;
(function (CacheSetPopResponse) {
    CacheSetPopResponse["Hit"] = "Hit";
    CacheSetPopResponse["Miss"] = "Miss";
    CacheSetPopResponse["Error"] = "Error";
})(CacheSetPopResponse = exports.CacheSetPopResponse || (exports.CacheSetPopResponse = {}));
var CacheSetLengthResponse;
(function (CacheSetLengthResponse) {
    CacheSetLengthResponse["Hit"] = "Hit";
    CacheSetLengthResponse["Miss"] = "Miss";
    CacheSetLengthResponse["Error"] = "Error";
})(CacheSetLengthResponse = exports.CacheSetLengthResponse || (exports.CacheSetLengthResponse = {}));
