// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/enums/cache/dictionary/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheDictionarySetFieldsResponse = exports.CacheDictionarySetFieldResponse = exports.CacheDictionaryRemoveFieldsResponse = exports.CacheDictionaryRemoveFieldResponse = exports.CacheDictionaryLengthResponse = exports.CacheDictionaryIncrementResponse = exports.CacheDictionaryFetchResponse = exports.CacheDictionaryGetFieldResponse = exports.CacheDictionaryGetFieldsResponse = void 0;
var CacheDictionaryGetFieldsResponse;
(function (CacheDictionaryGetFieldsResponse) {
    CacheDictionaryGetFieldsResponse["Hit"] = "Hit";
    CacheDictionaryGetFieldsResponse["Miss"] = "Miss";
    CacheDictionaryGetFieldsResponse["Error"] = "Error";
})(CacheDictionaryGetFieldsResponse = exports.CacheDictionaryGetFieldsResponse || (exports.CacheDictionaryGetFieldsResponse = {}));
var CacheDictionaryGetFieldResponse;
(function (CacheDictionaryGetFieldResponse) {
    CacheDictionaryGetFieldResponse["Hit"] = "Hit";
    CacheDictionaryGetFieldResponse["Miss"] = "Miss";
    CacheDictionaryGetFieldResponse["Error"] = "Error";
})(CacheDictionaryGetFieldResponse = exports.CacheDictionaryGetFieldResponse || (exports.CacheDictionaryGetFieldResponse = {}));
var CacheDictionaryFetchResponse;
(function (CacheDictionaryFetchResponse) {
    CacheDictionaryFetchResponse["Hit"] = "Hit";
    CacheDictionaryFetchResponse["Miss"] = "Miss";
    CacheDictionaryFetchResponse["Error"] = "Error";
})(CacheDictionaryFetchResponse = exports.CacheDictionaryFetchResponse || (exports.CacheDictionaryFetchResponse = {}));
var CacheDictionaryIncrementResponse;
(function (CacheDictionaryIncrementResponse) {
    CacheDictionaryIncrementResponse["Success"] = "Success";
    CacheDictionaryIncrementResponse["Error"] = "Error";
})(CacheDictionaryIncrementResponse = exports.CacheDictionaryIncrementResponse || (exports.CacheDictionaryIncrementResponse = {}));
var CacheDictionaryLengthResponse;
(function (CacheDictionaryLengthResponse) {
    CacheDictionaryLengthResponse["Hit"] = "Hit";
    CacheDictionaryLengthResponse["Miss"] = "Miss";
    CacheDictionaryLengthResponse["Error"] = "Error";
})(CacheDictionaryLengthResponse = exports.CacheDictionaryLengthResponse || (exports.CacheDictionaryLengthResponse = {}));
var CacheDictionaryRemoveFieldResponse;
(function (CacheDictionaryRemoveFieldResponse) {
    CacheDictionaryRemoveFieldResponse["Success"] = "Success";
    CacheDictionaryRemoveFieldResponse["Error"] = "Error";
})(CacheDictionaryRemoveFieldResponse = exports.CacheDictionaryRemoveFieldResponse || (exports.CacheDictionaryRemoveFieldResponse = {}));
var CacheDictionaryRemoveFieldsResponse;
(function (CacheDictionaryRemoveFieldsResponse) {
    CacheDictionaryRemoveFieldsResponse["Success"] = "Success";
    CacheDictionaryRemoveFieldsResponse["Error"] = "Error";
})(CacheDictionaryRemoveFieldsResponse = exports.CacheDictionaryRemoveFieldsResponse || (exports.CacheDictionaryRemoveFieldsResponse = {}));
var CacheDictionarySetFieldResponse;
(function (CacheDictionarySetFieldResponse) {
    CacheDictionarySetFieldResponse["Success"] = "Success";
    CacheDictionarySetFieldResponse["Error"] = "Error";
})(CacheDictionarySetFieldResponse = exports.CacheDictionarySetFieldResponse || (exports.CacheDictionarySetFieldResponse = {}));
var CacheDictionarySetFieldsResponse;
(function (CacheDictionarySetFieldsResponse) {
    CacheDictionarySetFieldsResponse["Success"] = "Success";
    CacheDictionarySetFieldsResponse["Error"] = "Error";
})(CacheDictionarySetFieldsResponse = exports.CacheDictionarySetFieldsResponse || (exports.CacheDictionarySetFieldsResponse = {}));
