// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/enums/cache/list/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheListPopFrontResponse = exports.CacheListPopBackResponse = exports.CacheListLengthResponse = exports.CacheListFetchResponse = exports.CacheListRetainResponse = exports.CacheListRemoveValueResponse = exports.CacheListPushFrontResponse = exports.CacheListPushBackResponse = exports.CacheListConcatenateFrontResponse = exports.CacheListConcatenateBackResponse = void 0;
var CacheListConcatenateBackResponse;
(function (CacheListConcatenateBackResponse) {
    CacheListConcatenateBackResponse["Success"] = "Success";
    CacheListConcatenateBackResponse["Error"] = "Error";
})(CacheListConcatenateBackResponse = exports.CacheListConcatenateBackResponse || (exports.CacheListConcatenateBackResponse = {}));
var CacheListConcatenateFrontResponse;
(function (CacheListConcatenateFrontResponse) {
    CacheListConcatenateFrontResponse["Success"] = "Success";
    CacheListConcatenateFrontResponse["Error"] = "Error";
})(CacheListConcatenateFrontResponse = exports.CacheListConcatenateFrontResponse || (exports.CacheListConcatenateFrontResponse = {}));
var CacheListPushBackResponse;
(function (CacheListPushBackResponse) {
    CacheListPushBackResponse["Success"] = "Success";
    CacheListPushBackResponse["Error"] = "Error";
})(CacheListPushBackResponse = exports.CacheListPushBackResponse || (exports.CacheListPushBackResponse = {}));
var CacheListPushFrontResponse;
(function (CacheListPushFrontResponse) {
    CacheListPushFrontResponse["Success"] = "Success";
    CacheListPushFrontResponse["Error"] = "Error";
})(CacheListPushFrontResponse = exports.CacheListPushFrontResponse || (exports.CacheListPushFrontResponse = {}));
var CacheListRemoveValueResponse;
(function (CacheListRemoveValueResponse) {
    CacheListRemoveValueResponse["Success"] = "Success";
    CacheListRemoveValueResponse["Error"] = "Error";
})(CacheListRemoveValueResponse = exports.CacheListRemoveValueResponse || (exports.CacheListRemoveValueResponse = {}));
var CacheListRetainResponse;
(function (CacheListRetainResponse) {
    CacheListRetainResponse["Success"] = "Success";
    CacheListRetainResponse["Error"] = "Error";
})(CacheListRetainResponse = exports.CacheListRetainResponse || (exports.CacheListRetainResponse = {}));
var CacheListFetchResponse;
(function (CacheListFetchResponse) {
    CacheListFetchResponse["Hit"] = "Hit";
    CacheListFetchResponse["Miss"] = "Miss";
    CacheListFetchResponse["Error"] = "Error";
})(CacheListFetchResponse = exports.CacheListFetchResponse || (exports.CacheListFetchResponse = {}));
var CacheListLengthResponse;
(function (CacheListLengthResponse) {
    CacheListLengthResponse["Hit"] = "Hit";
    CacheListLengthResponse["Miss"] = "Miss";
    CacheListLengthResponse["Error"] = "Error";
})(CacheListLengthResponse = exports.CacheListLengthResponse || (exports.CacheListLengthResponse = {}));
var CacheListPopBackResponse;
(function (CacheListPopBackResponse) {
    CacheListPopBackResponse["Hit"] = "Hit";
    CacheListPopBackResponse["Miss"] = "Miss";
    CacheListPopBackResponse["Error"] = "Error";
})(CacheListPopBackResponse = exports.CacheListPopBackResponse || (exports.CacheListPopBackResponse = {}));
var CacheListPopFrontResponse;
(function (CacheListPopFrontResponse) {
    CacheListPopFrontResponse["Hit"] = "Hit";
    CacheListPopFrontResponse["Miss"] = "Miss";
    CacheListPopFrontResponse["Error"] = "Error";
})(CacheListPopFrontResponse = exports.CacheListPopFrontResponse || (exports.CacheListPopFrontResponse = {}));
