// @gomomento/sdk@1.106.0/dist/src/config/transport/topics/transport-strategy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticTopicTransportStrategy = exports.StaticTopicGrpcConfiguration = void 0;
class StaticTopicGrpcConfiguration {
    constructor(props) {
        if (props.numClients !== undefined && props.numClients !== null) {
            this.numClients = props.numClients;
        }
        else {
            this.numClients = 1;
        }
        this.keepAliveTimeMs = props.keepAliveTimeMs;
        this.keepAliveTimeoutMs = props.keepAliveTimeoutMs;
        this.keepAlivePermitWithoutCalls = props.keepAlivePermitWithoutCalls;
    }
    getNumClients() {
        return this.numClients;
    }
    withNumClients(numClients) {
        return new StaticTopicGrpcConfiguration({
            ...this,
            numClients,
        });
    }
    getKeepAliveTimeoutMS() {
        return this.keepAliveTimeoutMs;
    }
    getKeepAliveTimeMS() {
        return this.keepAliveTimeMs;
    }
    getKeepAlivePermitWithoutCalls() {
        return this.keepAlivePermitWithoutCalls;
    }
}
exports.StaticTopicGrpcConfiguration = StaticTopicGrpcConfiguration;
class StaticTopicTransportStrategy {
    constructor(props) {
        this.grpcConfiguration = props.grpcConfiguration;
    }
    getGrpcConfig() {
        return this.grpcConfiguration;
    }
    withGrpcConfig(grpcConfiguration) {
        return new StaticTopicTransportStrategy({
            ...this,
            grpcConfiguration,
        });
    }
}
exports.StaticTopicTransportStrategy = StaticTopicTransportStrategy;
