// @gomomento/sdk@1.106.0/dist/src/internal/grpc/headers-interceptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeaderInterceptor = exports.Header = void 0;
const grpc_js_1 = require("@grpc/grpc-js");
class Header {
    /**
     * @param {string} name
     * @param {string} value
     */
    constructor(name, value) {
        this.onceOnlyHeaders = ['agent', 'runtime-version'];
        this.name = name;
        this.value = value;
    }
}
exports.Header = Header;
class HeaderInterceptor {
    static createHeadersInterceptor(headers) {
        const headersToAddOnce = headers.filter(header => header.onceOnlyHeaders.includes(header.name));
        const headersToAddEveryTime = headers.filter(header => !header.onceOnlyHeaders.includes(header.name));
        let areOnlyOnceHeadersSent = false;
        return (options, nextCall) => {
            return new grpc_js_1.InterceptingCall(nextCall(options), {
                start: (metadata, listener, next) => {
                    headersToAddEveryTime.forEach(h => {
                        metadata.set(h.name, h.value);
                    });
                    if (!areOnlyOnceHeadersSent) {
                        areOnlyOnceHeadersSent = true;
                        headersToAddOnce.forEach(h => metadata.add(h.name, h.value));
                    }
                    next(metadata, {});
                },
            });
        };
    }
}
exports.HeaderInterceptor = HeaderInterceptor;
