// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/leaderboard/leaderboard-fetch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = void 0;
const response_base_1 = require("../response-base");
const enums_1 = require("../enums");
/**
 * Indicates a Successful leaderboard fetch request.
 */
class Success extends response_base_1.BaseResponseSuccess {
    constructor(elements) {
        super();
        this.type = enums_1.LeaderboardFetchResponse.Success;
        this._elements = elements;
    }
    /**
     * Returns the elements as an array of objects, each containing an `id`, `score`, and `rank` field.
     * @returns {{id: number, score: number, rank: number}[]}
     */
    values() {
        return this._elements.map(item => {
            return {
                id: item.id,
                score: item.score,
                rank: item.rank,
            };
        });
    }
}
exports.Success = Success;
/**
 * Indicates that an error occurred during the leaderboard fetch by rank or by score request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.LeaderboardFetchResponse.Error;
    }
}
exports.Error = Error;
