// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-keys-exist.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
const TEXT_DECODER = new TextDecoder();
/**
 * Indicates a successful keys exist request.
 */
class Success extends response_base_1.BaseResponseSuccess {
    constructor(keys, exists) {
        super();
        this.type = enums_1.CacheKeysExistResponse.Success;
        this._keys = keys;
        this._exists = exists;
    }
    /**
     * A list of booleans indicating whether each given key was found in the cache.
     * @returns {boolean[]}
     */
    exists() {
        return this._exists;
    }
    /**
     * A record of key-exists pairs indicating whether each given key was found in the cache.
     * @returns {Record<string, boolean>}
     */
    valueRecord() {
        return this._keys.reduce((acc, field, index) => {
            acc[TEXT_DECODER.decode(field)] = this._exists[index];
            return acc;
        }, {});
    }
    toString() {
        const booleans = this._exists.map(bool => bool);
        return super.toString() + ': exists: ' + booleans.join(', ');
    }
}
exports.Success = Success;
/**
 * Indicates that an error occurred during the keys exist request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.CacheKeysExistResponse.Error;
    }
    exists() {
        return undefined;
    }
}
exports.Error = Error;
