// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-sorted-set-get-rank.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
/**
 * Indicates that the requested data was successfully retrieved from the cache.  Provides
 * `rank()` accessor to retrieve rank.
 */
class Hit extends response_base_1.ResponseBase {
    constructor(rank) {
        super();
        this.type = enums_1.CacheSortedSetGetRankResponse.Hit;
        this._rank = rank;
    }
    /**
     * Returns the rank of the element in the sorted set.  Ranks start at 0.
     * @returns {number}
     */
    rank() {
        return this._rank;
    }
    toString() {
        return `${super.toString()}: rank: ${this.rank()}`;
    }
}
exports.Hit = Hit;
/**
 * Indicates that the requested data was not available in the cache.
 */
class Miss extends response_base_1.BaseResponseMiss {
    constructor() {
        super();
        this.type = enums_1.CacheSortedSetGetRankResponse.Miss;
    }
    rank() {
        return undefined;
    }
}
exports.Miss = Miss;
/**
 * Indicates that an error occurred during the sorted set get rank request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(error) {
        super(error);
        this.type = enums_1.CacheSortedSetGetRankResponse.Error;
    }
    rank() {
        return undefined;
    }
}
exports.Error = Error;
