// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-set-contains-elements.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = void 0;
const response_base_1 = require("./response-base");
const utils_1 = require("../../internal/utils");
const enums_1 = require("./enums");
const TEXT_DECODER = new TextDecoder();
/**
 * Indicates that the requested set was in the cache.  Provides
 * a `containsElements` accessor that returns a mapping of the elements of the set
 * to their presence in the cache.
 */
class Hit extends response_base_1.ResponseBase {
    constructor(elements, found) {
        super();
        this.type = enums_1.CacheSetContainsElementsResponse.Hit;
        this._contains = elements.reduce((acc, element, index) => {
            acc[TEXT_DECODER.decode(element)] = found[index];
            return acc;
        }, {});
    }
    /**
     * Returns a mapping of the elements to their presence in the cache.
     * @returns {Record<string, boolean>}
     */
    containsElements() {
        return this._contains;
    }
    toString() {
        const display = (0, utils_1.truncateString)(Object.entries(this._contains)
            .map(([element, isPresent]) => `${element}: ${isPresent ? 'true' : 'false'}`)
            .join(', '));
        return `${super.toString()}: ${display}`;
    }
}
exports.Hit = Hit;
/**
 * Indicates that the requested set was not in the cache.
 */
class Miss extends response_base_1.BaseResponseMiss {
    constructor() {
        super();
        this.type = enums_1.CacheSetContainsElementsResponse.Miss;
    }
    containsElements() {
        return undefined;
    }
}
exports.Miss = Miss;
/**
 * Indicates that an error occurred during the set contains element request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.CacheSetContainsElementsResponse.Error;
    }
    containsElements() {
        return undefined;
    }
}
exports.Error = Error;
