// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-sorted-set-get-score.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
/**
 * Indicates that the requested data was successfully retrieved from the cache.  Provides
 * `score()` accessor to retrieve score.
 */
class Hit extends response_base_1.ResponseBase {
    constructor(score, value) {
        super();
        this.type = enums_1.CacheSortedSetGetScoreResponse.Hit;
        this._value = value;
        this._score = score;
    }
    /**
     * Returns the value of the element in the sorted set as a Uint8Array.
     * @returns {number}
     */
    valueUint8Array() {
        return this._value;
    }
    /**
     * Returns the value of the element in the sorted set as a string.
     * @returns {string}
     */
    valueString() {
        return new TextDecoder().decode(this._value);
    }
    /**
     * Returns the score of the element in the sorted set.
     * @return {*}  {number}
     */
    score() {
        return this._score;
    }
    toString() {
        return `${super.toString()}: value: ${this.valueString()}`;
    }
}
exports.Hit = Hit;
/**
 * Indicates that the requested data was not available in the cache.
 */
class Miss extends response_base_1.BaseResponseMiss {
    constructor(value) {
        super();
        this.type = enums_1.CacheSortedSetGetScoreResponse.Miss;
        this._value = value;
    }
    /**
     * Returns the value of the element in the sorted set as a Uint8Array.
     * @returns {number}
     */
    valueUint8Array() {
        return this._value;
    }
    /**
     * Returns the score of the element in the sorted set as a string.
     * @returns {string}
     */
    valueString() {
        return new TextDecoder().decode(this._value);
    }
    score() {
        return undefined;
    }
    toString() {
        return `${super.toString()}: value: ${this.valueString()}`;
    }
}
exports.Miss = Miss;
/**
 * Indicates that an error occurred during the sorted set get score request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException, value) {
        super(_innerException);
        this.type = enums_1.CacheSortedSetGetScoreResponse.Error;
        this._value = value;
    }
    /**
     * Returns the value of the element in the sorted set as a Uint8Array.
     * @returns {number}
     */
    valueUint8Array() {
        return this._value;
    }
    /**
     * Returns the score of the element in the sorted set as a string.
     * @returns {string}
     */
    valueString() {
        return new TextDecoder().decode(this._value);
    }
    score() {
        return undefined;
    }
}
exports.Error = Error;
