// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-item-get-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
/**
 * Indicates a successful item get type request.
 */
class Hit extends response_base_1.ResponseBase {
    constructor(keyType) {
        super();
        this.type = enums_1.CacheItemGetTypeResponse.Hit;
        this.keyType = keyType;
    }
    /**
     * Returns the type of key.
     * @returns {ItemType}
     */
    itemType() {
        return this.keyType;
    }
    toString() {
        return `${super.toString()}: item type: ${this.keyType}`;
    }
}
exports.Hit = Hit;
/**
 * Indicates that the requested item was not available in the cache.
 */
class Miss extends response_base_1.BaseResponseMiss {
    constructor() {
        super();
        this.type = enums_1.CacheItemGetTypeResponse.Miss;
    }
    itemType() {
        return undefined;
    }
}
exports.Miss = Miss;
/**
 * Indicates that an error occurred during the item get type request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.CacheItemGetTypeResponse.Error;
    }
    itemType() {
        return undefined;
    }
}
exports.Error = Error;
