// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-sorted-set-length-by-score.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
/**
 * Indicates that the requested data was successfully retrieved from the cache.  Provides
 * `value*` accessors to retrieve the data in the appropriate format.
 */
class Hit extends response_base_1.ResponseBase {
    constructor(length) {
        super();
        this.type = enums_1.CacheSortedSetLengthByScoreResponse.Hit;
        this._length = length;
    }
    /**
     * Returns the length of the sorted set
     * @returns {number}
     */
    length() {
        return this._length;
    }
    toString() {
        return `${super.toString()}: length ${this._length}`;
    }
}
exports.Hit = Hit;
/**
 * Indicates that the requested data was not available in the cache.
 */
class Miss extends response_base_1.BaseResponseMiss {
    constructor() {
        super(...arguments);
        this.type = enums_1.CacheSortedSetLengthByScoreResponse.Miss;
    }
    length() {
        return undefined;
    }
}
exports.Miss = Miss;
/**
 * Indicates that an error occurred during the sorted set length request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(error) {
        super(error);
        this.type = enums_1.CacheSortedSetLengthByScoreResponse.Error;
    }
    length() {
        return undefined;
    }
}
exports.Error = Error;
