// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-set-contains-element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
/**
 * Indicates that the requested set was in the cache.
 * Provides a `containsElement` accessor that returns a boolean indicating whether the element was found.
 */
class Hit extends response_base_1.ResponseBase {
    constructor(found) {
        super();
        this.type = enums_1.CacheSetContainsElementResponse.Hit;
        this._containsElement = found;
    }
    /**
     * Returns a boolean indicating whether the element was found in the cache.
     * @returns {boolean}
     */
    containsElement() {
        return this._containsElement;
    }
    toString() {
        return `${super.toString()}: Hit - ${this._containsElement ? 'true' : 'false'}`;
    }
}
exports.Hit = Hit;
/**
 * Indicates that the requested set was not available in the cache.
 */
class Miss extends response_base_1.BaseResponseMiss {
    constructor() {
        super();
        this.type = enums_1.CacheSetContainsElementResponse.Miss;
    }
    containsElement() {
        return undefined;
    }
}
exports.Miss = Miss;
/**
 * Indicates that an error occurred during the set contains element request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.CacheSetContainsElementResponse.Error;
    }
    containsElement() {
        return undefined;
    }
}
exports.Error = Error;
