// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-increment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
/**
 * Indicates a successful cache increment request.
 * Provides `value*` accessors to retrieve the updated value in the appropriate format.
 */
class Success extends response_base_1.BaseResponseSuccess {
    constructor(body) {
        super();
        this.type = enums_1.CacheIncrementResponse.Success;
        this.body = body;
    }
    /**
     * Returns the updated value as a number.
     * @returns number
     */
    value() {
        return this.body;
    }
    /**
     * Returns the updated value as a number.
     * @returns number
     */
    valueNumber() {
        return this.body;
    }
    toString() {
        return `${super.toString()}: value: ${this.value()}`;
    }
}
exports.Success = Success;
/**
 * Indicates that an error occurred during the increment request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error.
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.CacheIncrementResponse.Error;
    }
    value() {
        return undefined;
    }
}
exports.Error = Error;
