// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-list-fetch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = void 0;
const response_base_1 = require("./response-base");
const utils_1 = require("../../internal/utils");
const enums_1 = require("./enums");
const TEXT_DECODER = new TextDecoder();
/**
 * Indicates that the requested list was successfully retrieved from the cache.
 * Provides `value*` accessors to retrieve the data in the appropriate format.
 */
class Hit extends response_base_1.ResponseBase {
    constructor(values) {
        super();
        this.type = enums_1.CacheListFetchResponse.Hit;
        this._values = values;
    }
    /**
     * Returns the data as an array of byte arrays.
     * @returns {Uint8Array[]}
     */
    valueListUint8Array() {
        return this._values;
    }
    /**
     * Returns the data as an array of strings, decoded from the underlying byte array.
     * @returns {string[]}
     */
    valueListString() {
        return this._values.map(v => TEXT_DECODER.decode(v));
    }
    /**
     * Returns the data as an array of strings, decoded from the underlying byte array.  This is a convenience alias
     * for {valueListString}
     * @returns {string[]}
     */
    valueList() {
        return this.valueListString();
    }
    /**
     * Returns the data as an array of strings, decoded from the underlying byte array.
     * This is a convenience alias for {valueListString}
     * @returns {string[]}
     */
    value() {
        return this.valueListString();
    }
    toString() {
        const truncatedStringArray = (0, utils_1.truncateStringArray)(this.valueListString());
        return `${super.toString()}: [${truncatedStringArray.toString()}]`;
    }
}
exports.Hit = Hit;
/**
 * Indicates that the requested list was not available in the cache.
 */
class Miss extends response_base_1.BaseResponseMiss {
    constructor() {
        super();
        this.type = enums_1.CacheListFetchResponse.Miss;
    }
    value() {
        return undefined;
    }
}
exports.Miss = Miss;
/**
 * Indicates that an error occurred during the list fetch request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.CacheListFetchResponse.Error;
    }
    value() {
        return undefined;
    }
}
exports.Error = Error;
