// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/cache-set-fetch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = void 0;
const response_base_1 = require("./response-base");
const utils_1 = require("../../internal/utils");
const enums_1 = require("./enums");
const TEXT_DECODER = new TextDecoder();
/**
 * Indicates that the requested data was successfully retrieved from the cache.  Provides
 * `value*` accessors to retrieve the data in the appropriate format.
 */
class Hit extends response_base_1.ResponseBase {
    constructor(elements) {
        super();
        this.type = enums_1.CacheSetFetchResponse.Hit;
        this.elements = elements;
    }
    /**
     * Returns the data as a Set whose values are utf-8 strings, decoded from the underlying byte arrays.  This
     * is a convenience alias for {valueSetString}.
     * @returns {Set<string>}
     */
    valueSet() {
        return this.valueSetString();
    }
    /**
     * Returns the data as a Set whose values are utf-8 strings, decoded from the underlying byte arrays.
     * @returns {Set<string>}
     */
    valueSetString() {
        return new Set(this.elements.map(e => TEXT_DECODER.decode(e)));
    }
    /**
     * Returns the data as a Set whose values are byte arrays.
     * @returns {Set<Uint8Array>}
     */
    valueSetUint8Array() {
        return new Set(this.elements);
    }
    /**
     * Returns the data as an Array whose values are utf-8 strings, decoded from the underlying byte arrays.
     * This accessor is provided because Arrays are sometimes easier to work with in TypeScript/JavaScript than Sets are.
     * This is a convenience alias for {valueArrayString}.
     * @returns {string[]}
     */
    value() {
        return this.valueArrayString();
    }
    /**
     * Returns the data as an Array whose values are utf-8 strings, decoded from the underlying byte arrays.
     * This accessor is provided because Arrays are sometimes easier to work with in TypeScript/JavaScript than Sets are.
     * This is a convenience alias for {valueArrayString}.
     * @returns {string[]}
     */
    valueArray() {
        return this.valueArrayString();
    }
    /**
     * Returns the data as an Array whose values are utf-8 strings, decoded from the underlying byte arrays.
     * This accessor is provided because Arrays are sometimes easier to work with in TypeScript/JavaScript than Sets are.
     * @returns {string[]}
     */
    valueArrayString() {
        return this.elements.map(e => TEXT_DECODER.decode(e));
    }
    /**
     * Returns the data as an Array whose values are byte arrays.
     * This accessor is provided because Arrays are sometimes easier to work with in TypeScript/JavaScript than Sets are.
     * @returns {Uint8Array[]}
     */
    valueArrayUint8Array() {
        return this.elements;
    }
    toString() {
        const truncatedStringArray = (0, utils_1.truncateStringArray)(Array.from(this.valueSetString()));
        return `${super.toString()}: [${truncatedStringArray.toString()}]`;
    }
}
exports.Hit = Hit;
/**
 * Indicates that the requested data was not available in the cache.
 */
class Miss extends response_base_1.BaseResponseMiss {
    constructor() {
        super(...arguments);
        this.type = enums_1.CacheSetFetchResponse.Miss;
    }
    value() {
        return undefined;
    }
}
exports.Miss = Miss;
/**
 * Indicates that an error occurred during the set fetch request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.CacheSetFetchResponse.Error;
    }
    value() {
        return undefined;
    }
}
exports.Error = Error;
