// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/generate-disposable-token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
const utils_1 = require("../../internal/utils");
class Success extends response_base_1.BaseResponseSuccess {
    constructor(apiKey, endpoint, expiresAt) {
        super();
        this.type = enums_1.GenerateDisposableTokenResponse.Success;
        this.authToken = (0, utils_1.encodeToBase64)(JSON.stringify({ endpoint: endpoint, api_key: apiKey }));
        this.endpoint = endpoint;
        this.expiresAt = expiresAt;
    }
}
exports.Success = Success;
/**
 * Indicates that an error occurred during the generate disposable token request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(innerException) {
        super(innerException);
        this.type = enums_1.GenerateDisposableTokenResponse.Error;
    }
}
exports.Error = Error;
