// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/list-caches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
/**
 * Indicates a successful list caches request.
 */
class Success extends response_base_1.BaseResponseSuccess {
    constructor(caches) {
        super();
        this.type = enums_1.ListCachesResponse.Success;
        this.caches = caches;
    }
    /**
     * An array of CacheInfo, containing information about each cache.
     * @returns {CacheInfo[]}
     */
    getCaches() {
        return this.caches;
    }
    toString() {
        const caches = this.caches.map(cacheInfo => cacheInfo.getName());
        return super.toString() + ': ' + caches.join(', ');
    }
}
exports.Success = Success;
/**
 * Indicates that an error occurred during the list caches request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.ListCachesResponse.Error;
    }
}
exports.Error = Error;
