// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/response-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseNotFound = exports.ResponseFound = exports.ResponseSuccess = exports.BaseResponseSubscription = exports.BaseResponseSuccess = exports.ResponseMiss = exports.BaseResponseMiss = exports.ResponseHit = exports.ResponseError = exports.BaseResponseError = exports.ResponseBase = void 0;
class ResponseBase {
    toString() {
        return this.constructor.name;
    }
}
exports.ResponseBase = ResponseBase;
class BaseResponseError extends ResponseBase {
    constructor(innerException) {
        super();
        this._innerException = innerException;
    }
    message() {
        return this._innerException.wrappedErrorMessage();
    }
    innerException() {
        return this._innerException;
    }
    errorCode() {
        return this._innerException.errorCode();
    }
    toString() {
        return this.message();
    }
}
exports.BaseResponseError = BaseResponseError;
function ResponseError(Base) {
    return class ResponseError extends Base {
        message() {
            return this._innerException.wrappedErrorMessage();
        }
        innerException() {
            return this._innerException;
        }
        errorCode() {
            return this._innerException.errorCode();
        }
        toString() {
            return this.message();
        }
    };
}
exports.ResponseError = ResponseError;
function ResponseHit(Base) {
    return class ResponseHit extends Base {
    };
}
exports.ResponseHit = ResponseHit;
class BaseResponseMiss extends ResponseBase {
    constructor() {
        super(...arguments);
        this.is_miss = true;
    }
}
exports.BaseResponseMiss = BaseResponseMiss;
function ResponseMiss(Base) {
    return class ResponseMiss extends Base {
        constructor() {
            super(...arguments);
            this.is_miss = true;
        }
    };
}
exports.ResponseMiss = ResponseMiss;
class BaseResponseSuccess extends ResponseBase {
    constructor() {
        super(...arguments);
        this.is_success = true;
    }
}
exports.BaseResponseSuccess = BaseResponseSuccess;
class BaseResponseSubscription extends ResponseBase {
    constructor() {
        super(...arguments);
        this.is_subscribed = true;
    }
}
exports.BaseResponseSubscription = BaseResponseSubscription;
function ResponseSuccess(Base) {
    return class ResponseSuccess extends Base {
        constructor() {
            super(...arguments);
            this.is_success = true;
        }
    };
}
exports.ResponseSuccess = ResponseSuccess;
function ResponseFound(Base) {
    return class ResponseFound extends Base {
    };
}
exports.ResponseFound = ResponseFound;
function ResponseNotFound(Base) {
    return class ResponseNotFound extends Base {
    };
}
exports.ResponseNotFound = ResponseNotFound;
