// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/topic-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopicItem = void 0;
const utils_1 = require("../../internal/utils");
/**
 * Represents the data received from a topic subscription.
 *
 * @remarks A subscription is created by calling {@link TopicClient.subscribe}.
 * The value is guaranteed to be either a {@link string} or a {@link Uint8Array}.
 * Call the appropriate accessor if you know the type of the value.
 */
class TopicItem {
    constructor(_value, _sequenceNumber, options) {
        this._value = _value;
        this._tokenId = options === null || options === void 0 ? void 0 : options.tokenId;
        this._sequenceNumber = _sequenceNumber;
    }
    /**
     * Returns the data read from the stream.
     * @returns string | Uint8Array
     */
    value() {
        return this._value;
    }
    /**
     * Returns the data read from the stream as a {@link string}.
     * @returns string
     */
    valueString() {
        return this.value().toString();
    }
    /**
     * Returns the data read from the stream as a {@link Uint8Array}.
     * @returns Uint8Array
     */
    valueUint8Array() {
        return this.value();
    }
    /**
     * Optionally returns the publisher ID from the steam if it exists.
     * @returns string | undefined
     */
    tokenId() {
        return this._tokenId;
    }
    /**
     * Returns the sequence number of the item.
     * @returns number
     */
    sequenceNumber() {
        return this._sequenceNumber;
    }
    toString() {
        const displayValue = (0, utils_1.truncateString)(this.value().toString());
        let displayString = `${this.constructor.name}: ${displayValue}`;
        if (this._tokenId !== undefined) {
            displayString += `; Token Id: ${this._tokenId}`;
        }
        return displayString;
    }
}
exports.TopicItem = TopicItem;
