// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/topic-subscribe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Subscription = void 0;
const response_base_1 = require("./response-base");
const enums_1 = require("./enums");
/**
 * Encapsulates a topic subscription.
 *
 * @remarks Currently allows unsubscribing from the topic.
 * In the future, this may be extended to include additional
 * statistics about the subscription.
 */
class Subscription extends response_base_1.BaseResponseSubscription {
    constructor(loggerFactory, subscriptionState) {
        super();
        this.type = enums_1.TopicSubscribeResponse.Subscription;
        this.logger = loggerFactory.getLogger(this);
        this.subscriptionState = subscriptionState;
    }
    /**
     * Unsubscribes from the topic.
     *
     * @returns void
     */
    unsubscribe() {
        this.logger.trace(`Unsubscribing from subscription: ${this.subscriptionState.toString()}`);
        this.subscriptionState.unsubscribe();
    }
    get isSubscribed() {
        return this.subscriptionState.isSubscribed;
    }
}
exports.Subscription = Subscription;
/**
 * Indicates that an error occurred during the topic subscribe request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.TopicSubscribeResponse.Error;
    }
}
exports.Error = Error;
