// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/enums/webhook/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListWebhooksResponse = exports.GetWebhookSecretResponse = exports.RotateWebhookSecretResponse = exports.DeleteWebhookResponse = exports.PutWebhookResponse = void 0;
var PutWebhookResponse;
(function (PutWebhookResponse) {
    PutWebhookResponse["Success"] = "Success";
    PutWebhookResponse["Error"] = "Error";
})(PutWebhookResponse = exports.PutWebhookResponse || (exports.PutWebhookResponse = {}));
var DeleteWebhookResponse;
(function (DeleteWebhookResponse) {
    DeleteWebhookResponse["Success"] = "Success";
    DeleteWebhookResponse["Error"] = "Error";
})(DeleteWebhookResponse = exports.DeleteWebhookResponse || (exports.DeleteWebhookResponse = {}));
var RotateWebhookSecretResponse;
(function (RotateWebhookSecretResponse) {
    RotateWebhookSecretResponse["Success"] = "Success";
    RotateWebhookSecretResponse["Error"] = "Error";
})(RotateWebhookSecretResponse = exports.RotateWebhookSecretResponse || (exports.RotateWebhookSecretResponse = {}));
var GetWebhookSecretResponse;
(function (GetWebhookSecretResponse) {
    GetWebhookSecretResponse["Success"] = "Success";
    GetWebhookSecretResponse["Error"] = "Error";
})(GetWebhookSecretResponse = exports.GetWebhookSecretResponse || (exports.GetWebhookSecretResponse = {}));
var ListWebhooksResponse;
(function (ListWebhooksResponse) {
    ListWebhooksResponse["Success"] = "Success";
    ListWebhooksResponse["Error"] = "Error";
})(ListWebhooksResponse = exports.ListWebhooksResponse || (exports.ListWebhooksResponse = {}));
