// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/enums/cache/sorted-set/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheSortedSetRemoveElementsResponse = exports.CacheSortedSetRemoveElementResponse = exports.CacheSortedSetPutElementsResponse = exports.CacheSortedSetPutElementResponse = exports.CacheSortedSetLengthByScoreResponse = exports.CacheSortedSetLengthResponse = exports.CacheSortedSetIncrementScoreResponse = exports.CacheSortedSetGetScoresResponse = exports.CacheSortedSetGetScoreResponse = exports.CacheSortedSetGetRankResponse = exports.CacheSortedSetFetchResponse = void 0;
var CacheSortedSetFetchResponse;
(function (CacheSortedSetFetchResponse) {
    CacheSortedSetFetchResponse["Hit"] = "Hit";
    CacheSortedSetFetchResponse["Miss"] = "Miss";
    CacheSortedSetFetchResponse["Error"] = "Error";
})(CacheSortedSetFetchResponse = exports.CacheSortedSetFetchResponse || (exports.CacheSortedSetFetchResponse = {}));
var CacheSortedSetGetRankResponse;
(function (CacheSortedSetGetRankResponse) {
    CacheSortedSetGetRankResponse["Hit"] = "Hit";
    CacheSortedSetGetRankResponse["Miss"] = "Miss";
    CacheSortedSetGetRankResponse["Error"] = "Error";
})(CacheSortedSetGetRankResponse = exports.CacheSortedSetGetRankResponse || (exports.CacheSortedSetGetRankResponse = {}));
var CacheSortedSetGetScoreResponse;
(function (CacheSortedSetGetScoreResponse) {
    CacheSortedSetGetScoreResponse["Hit"] = "Hit";
    CacheSortedSetGetScoreResponse["Miss"] = "Miss";
    CacheSortedSetGetScoreResponse["Error"] = "Error";
})(CacheSortedSetGetScoreResponse = exports.CacheSortedSetGetScoreResponse || (exports.CacheSortedSetGetScoreResponse = {}));
var CacheSortedSetGetScoresResponse;
(function (CacheSortedSetGetScoresResponse) {
    CacheSortedSetGetScoresResponse["Hit"] = "Hit";
    CacheSortedSetGetScoresResponse["Miss"] = "Miss";
    CacheSortedSetGetScoresResponse["Error"] = "Error";
})(CacheSortedSetGetScoresResponse = exports.CacheSortedSetGetScoresResponse || (exports.CacheSortedSetGetScoresResponse = {}));
var CacheSortedSetIncrementScoreResponse;
(function (CacheSortedSetIncrementScoreResponse) {
    CacheSortedSetIncrementScoreResponse["Success"] = "Success";
    CacheSortedSetIncrementScoreResponse["Error"] = "Error";
})(CacheSortedSetIncrementScoreResponse = exports.CacheSortedSetIncrementScoreResponse || (exports.CacheSortedSetIncrementScoreResponse = {}));
var CacheSortedSetLengthResponse;
(function (CacheSortedSetLengthResponse) {
    CacheSortedSetLengthResponse["Hit"] = "Hit";
    CacheSortedSetLengthResponse["Miss"] = "Miss";
    CacheSortedSetLengthResponse["Error"] = "Error";
})(CacheSortedSetLengthResponse = exports.CacheSortedSetLengthResponse || (exports.CacheSortedSetLengthResponse = {}));
var CacheSortedSetLengthByScoreResponse;
(function (CacheSortedSetLengthByScoreResponse) {
    CacheSortedSetLengthByScoreResponse["Hit"] = "Hit";
    CacheSortedSetLengthByScoreResponse["Miss"] = "Miss";
    CacheSortedSetLengthByScoreResponse["Error"] = "Error";
})(CacheSortedSetLengthByScoreResponse = exports.CacheSortedSetLengthByScoreResponse || (exports.CacheSortedSetLengthByScoreResponse = {}));
var CacheSortedSetPutElementResponse;
(function (CacheSortedSetPutElementResponse) {
    CacheSortedSetPutElementResponse["Success"] = "Success";
    CacheSortedSetPutElementResponse["Error"] = "Error";
})(CacheSortedSetPutElementResponse = exports.CacheSortedSetPutElementResponse || (exports.CacheSortedSetPutElementResponse = {}));
var CacheSortedSetPutElementsResponse;
(function (CacheSortedSetPutElementsResponse) {
    CacheSortedSetPutElementsResponse["Success"] = "Success";
    CacheSortedSetPutElementsResponse["Error"] = "Error";
})(CacheSortedSetPutElementsResponse = exports.CacheSortedSetPutElementsResponse || (exports.CacheSortedSetPutElementsResponse = {}));
var CacheSortedSetRemoveElementResponse;
(function (CacheSortedSetRemoveElementResponse) {
    CacheSortedSetRemoveElementResponse["Success"] = "Success";
    CacheSortedSetRemoveElementResponse["Error"] = "Error";
})(CacheSortedSetRemoveElementResponse = exports.CacheSortedSetRemoveElementResponse || (exports.CacheSortedSetRemoveElementResponse = {}));
var CacheSortedSetRemoveElementsResponse;
(function (CacheSortedSetRemoveElementsResponse) {
    CacheSortedSetRemoveElementsResponse["Success"] = "Success";
    CacheSortedSetRemoveElementsResponse["Error"] = "Error";
})(CacheSortedSetRemoveElementsResponse = exports.CacheSortedSetRemoveElementsResponse || (exports.CacheSortedSetRemoveElementsResponse = {}));
