// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/webhook/get-webhook-secret.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = void 0;
const response_base_1 = require("../response-base");
const enums_1 = require("../enums");
/**
 * Indicates a Successful GetWebhookSecret request.
 */
class Success extends response_base_1.BaseResponseSuccess {
    constructor(props) {
        super();
        this.type = enums_1.GetWebhookSecretResponse.Success;
        this._secret = props.secret;
        this._cacheName = props.cacheName;
        this._webhookName = props.webhookName;
    }
    secret() {
        return this._secret;
    }
    webhookName() {
        return this._webhookName;
    }
    cacheName() {
        return this._cacheName;
    }
    toString() {
        return super.toString() + ': ' + this._webhookName;
    }
}
exports.Success = Success;
/**
 * Indicates that an error occurred during the GetWebhookSecret request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.GetWebhookSecretResponse.Error;
    }
}
exports.Error = Error;
