// @gomomento/generated-types@0.119.2/dist/store.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.store = void 0;
/**
 * Generated by the protoc-gen-ts.  DO NOT EDIT!
 * compiler version: 3.19.1
 * source: store.proto
 * git: https://github.com/thesayyn/protoc-gen-ts */
const pb_1 = __importStar(require("google-protobuf"));
const grpc_1 = __importStar(require("@grpc/grpc-js"));
var store;
(function (store) {
    var __StoreValue_one_of_decls, __StoreGetRequest_one_of_decls, __StoreGetResponse_one_of_decls, __StorePutRequest_one_of_decls, __StorePutResponse_one_of_decls, __StoreDeleteRequest_one_of_decls, __StoreDeleteResponse_one_of_decls;
    class _StoreValue extends pb_1.Message {
        constructor(data) {
            super();
            __StoreValue_one_of_decls.set(this, [[1, 2, 3, 4]]);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], __classPrivateFieldGet(this, __StoreValue_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("bytes_value" in data && data.bytes_value != undefined) {
                    this.bytes_value = data.bytes_value;
                }
                if ("string_value" in data && data.string_value != undefined) {
                    this.string_value = data.string_value;
                }
                if ("integer_value" in data && data.integer_value != undefined) {
                    this.integer_value = data.integer_value;
                }
                if ("double_value" in data && data.double_value != undefined) {
                    this.double_value = data.double_value;
                }
            }
        }
        get bytes_value() {
            return pb_1.Message.getFieldWithDefault(this, 1, new Uint8Array(0));
        }
        set bytes_value(value) {
            pb_1.Message.setOneofField(this, 1, __classPrivateFieldGet(this, __StoreValue_one_of_decls, "f")[0], value);
        }
        get has_bytes_value() {
            return pb_1.Message.getField(this, 1) != null;
        }
        get string_value() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set string_value(value) {
            pb_1.Message.setOneofField(this, 2, __classPrivateFieldGet(this, __StoreValue_one_of_decls, "f")[0], value);
        }
        get has_string_value() {
            return pb_1.Message.getField(this, 2) != null;
        }
        get integer_value() {
            return pb_1.Message.getFieldWithDefault(this, 3, 0);
        }
        set integer_value(value) {
            pb_1.Message.setOneofField(this, 3, __classPrivateFieldGet(this, __StoreValue_one_of_decls, "f")[0], value);
        }
        get has_integer_value() {
            return pb_1.Message.getField(this, 3) != null;
        }
        get double_value() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set double_value(value) {
            pb_1.Message.setOneofField(this, 4, __classPrivateFieldGet(this, __StoreValue_one_of_decls, "f")[0], value);
        }
        get has_double_value() {
            return pb_1.Message.getField(this, 4) != null;
        }
        get value() {
            const cases = {
                0: "none",
                1: "bytes_value",
                2: "string_value",
                3: "integer_value",
                4: "double_value"
            };
            return cases[pb_1.Message.computeOneofCase(this, [1, 2, 3, 4])];
        }
        static fromObject(data) {
            const message = new _StoreValue({});
            if (data.bytes_value != null) {
                message.bytes_value = data.bytes_value;
            }
            if (data.string_value != null) {
                message.string_value = data.string_value;
            }
            if (data.integer_value != null) {
                message.integer_value = data.integer_value;
            }
            if (data.double_value != null) {
                message.double_value = data.double_value;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.bytes_value != null) {
                data.bytes_value = this.bytes_value;
            }
            if (this.string_value != null) {
                data.string_value = this.string_value;
            }
            if (this.integer_value != null) {
                data.integer_value = this.integer_value;
            }
            if (this.double_value != null) {
                data.double_value = this.double_value;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.has_bytes_value)
                writer.writeBytes(1, this.bytes_value);
            if (this.has_string_value)
                writer.writeString(2, this.string_value);
            if (this.has_integer_value)
                writer.writeInt64(3, this.integer_value);
            if (this.has_double_value)
                writer.writeDouble(4, this.double_value);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new _StoreValue();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.bytes_value = reader.readBytes();
                        break;
                    case 2:
                        message.string_value = reader.readString();
                        break;
                    case 3:
                        message.integer_value = reader.readInt64();
                        break;
                    case 4:
                        message.double_value = reader.readDouble();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return _StoreValue.deserialize(bytes);
        }
    }
    __StoreValue_one_of_decls = new WeakMap();
    store._StoreValue = _StoreValue;
    class _StoreGetRequest extends pb_1.Message {
        constructor(data) {
            super();
            __StoreGetRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], __classPrivateFieldGet(this, __StoreGetRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("key" in data && data.key != undefined) {
                    this.key = data.key;
                }
            }
        }
        get key() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set key(value) {
            pb_1.Message.setField(this, 1, value);
        }
        static fromObject(data) {
            const message = new _StoreGetRequest({});
            if (data.key != null) {
                message.key = data.key;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.key != null) {
                data.key = this.key;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.key.length)
                writer.writeString(1, this.key);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new _StoreGetRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.key = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return _StoreGetRequest.deserialize(bytes);
        }
    }
    __StoreGetRequest_one_of_decls = new WeakMap();
    store._StoreGetRequest = _StoreGetRequest;
    class _StoreGetResponse extends pb_1.Message {
        constructor(data) {
            super();
            __StoreGetResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], __classPrivateFieldGet(this, __StoreGetResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("value" in data && data.value != undefined) {
                    this.value = data.value;
                }
            }
        }
        get value() {
            return pb_1.Message.getWrapperField(this, _StoreValue, 1);
        }
        set value(value) {
            pb_1.Message.setWrapperField(this, 1, value);
        }
        get has_value() {
            return pb_1.Message.getField(this, 1) != null;
        }
        static fromObject(data) {
            const message = new _StoreGetResponse({});
            if (data.value != null) {
                message.value = _StoreValue.fromObject(data.value);
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.value != null) {
                data.value = this.value.toObject();
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.has_value)
                writer.writeMessage(1, this.value, () => this.value.serialize(writer));
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new _StoreGetResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        reader.readMessage(message.value, () => message.value = _StoreValue.deserialize(reader));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return _StoreGetResponse.deserialize(bytes);
        }
    }
    __StoreGetResponse_one_of_decls = new WeakMap();
    store._StoreGetResponse = _StoreGetResponse;
    class _StorePutRequest extends pb_1.Message {
        constructor(data) {
            super();
            __StorePutRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], __classPrivateFieldGet(this, __StorePutRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("key" in data && data.key != undefined) {
                    this.key = data.key;
                }
                if ("value" in data && data.value != undefined) {
                    this.value = data.value;
                }
            }
        }
        get key() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set key(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get value() {
            return pb_1.Message.getWrapperField(this, _StoreValue, 2);
        }
        set value(value) {
            pb_1.Message.setWrapperField(this, 2, value);
        }
        get has_value() {
            return pb_1.Message.getField(this, 2) != null;
        }
        static fromObject(data) {
            const message = new _StorePutRequest({});
            if (data.key != null) {
                message.key = data.key;
            }
            if (data.value != null) {
                message.value = _StoreValue.fromObject(data.value);
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.key != null) {
                data.key = this.key;
            }
            if (this.value != null) {
                data.value = this.value.toObject();
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.key.length)
                writer.writeString(1, this.key);
            if (this.has_value)
                writer.writeMessage(2, this.value, () => this.value.serialize(writer));
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new _StorePutRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.key = reader.readString();
                        break;
                    case 2:
                        reader.readMessage(message.value, () => message.value = _StoreValue.deserialize(reader));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return _StorePutRequest.deserialize(bytes);
        }
    }
    __StorePutRequest_one_of_decls = new WeakMap();
    store._StorePutRequest = _StorePutRequest;
    class _StorePutResponse extends pb_1.Message {
        constructor(data) {
            super();
            __StorePutResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], __classPrivateFieldGet(this, __StorePutResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") { }
        }
        static fromObject(data) {
            const message = new _StorePutResponse({});
            return message;
        }
        toObject() {
            const data = {};
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new _StorePutResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return _StorePutResponse.deserialize(bytes);
        }
    }
    __StorePutResponse_one_of_decls = new WeakMap();
    store._StorePutResponse = _StorePutResponse;
    class _StoreDeleteRequest extends pb_1.Message {
        constructor(data) {
            super();
            __StoreDeleteRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], __classPrivateFieldGet(this, __StoreDeleteRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("key" in data && data.key != undefined) {
                    this.key = data.key;
                }
            }
        }
        get key() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set key(value) {
            pb_1.Message.setField(this, 1, value);
        }
        static fromObject(data) {
            const message = new _StoreDeleteRequest({});
            if (data.key != null) {
                message.key = data.key;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.key != null) {
                data.key = this.key;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.key.length)
                writer.writeString(1, this.key);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new _StoreDeleteRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.key = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return _StoreDeleteRequest.deserialize(bytes);
        }
    }
    __StoreDeleteRequest_one_of_decls = new WeakMap();
    store._StoreDeleteRequest = _StoreDeleteRequest;
    class _StoreDeleteResponse extends pb_1.Message {
        constructor(data) {
            super();
            __StoreDeleteResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], __classPrivateFieldGet(this, __StoreDeleteResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") { }
        }
        static fromObject(data) {
            const message = new _StoreDeleteResponse({});
            return message;
        }
        toObject() {
            const data = {};
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new _StoreDeleteResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return _StoreDeleteResponse.deserialize(bytes);
        }
    }
    __StoreDeleteResponse_one_of_decls = new WeakMap();
    store._StoreDeleteResponse = _StoreDeleteResponse;
    class UnimplementedStoreService {
    }
    UnimplementedStoreService.definition = {
        Get: {
            path: "/store.Store/Get",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => _StoreGetRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => _StoreGetResponse.deserialize(new Uint8Array(bytes))
        },
        Put: {
            path: "/store.Store/Put",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => _StorePutRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => _StorePutResponse.deserialize(new Uint8Array(bytes))
        },
        Delete: {
            path: "/store.Store/Delete",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => _StoreDeleteRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => _StoreDeleteResponse.deserialize(new Uint8Array(bytes))
        }
    };
    store.UnimplementedStoreService = UnimplementedStoreService;
    class StoreClient extends grpc_1.makeGenericClientConstructor(UnimplementedStoreService.definition, "Store", {}) {
        constructor(address, credentials, options) {
            super(address, credentials, options);
            this.Get = (message, metadata, options, callback) => {
                return super.Get(message, metadata, options, callback);
            };
            this.Put = (message, metadata, options, callback) => {
                return super.Put(message, metadata, options, callback);
            };
            this.Delete = (message, metadata, options, callback) => {
                return super.Delete(message, metadata, options, callback);
            };
        }
    }
    store.StoreClient = StoreClient;
})(store = exports.store || (exports.store = {}));
