// @gomomento/sdk-core@1.106.0/dist/src/internal/subscription-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionState = void 0;
/**
 * Encapsulates a topic subscription stream state.
 */
class SubscriptionState {
    constructor() {
        this._unsubscribeFn = () => {
            return;
        };
        this._isSubscribed = false;
    }
    get resumeAtTopicSequenceNumber() {
        var _a;
        return ((_a = this.lastTopicSequenceNumber) !== null && _a !== void 0 ? _a : -1) + 1;
    }
    get resumeAtTopicSequencePage() {
        var _a;
        return (_a = this.lastTopicSequencePage) !== null && _a !== void 0 ? _a : 0;
    }
    setSubscribed() {
        this._isSubscribed = true;
    }
    setUnsubscribed() {
        this._isSubscribed = false;
    }
    get isSubscribed() {
        return this._isSubscribed;
    }
    set unsubscribeFn(unsubscribeFn) {
        this._unsubscribeFn = unsubscribeFn;
    }
    unsubscribe() {
        if (this.isSubscribed) {
            this._unsubscribeFn();
            this.setUnsubscribed();
        }
    }
    toString() {
        return JSON.stringify({
            lastTopicSequenceNumber: this.lastTopicSequenceNumber,
            lastTopicSequencePage: this.lastTopicSequencePage,
            isSubscribed: this._isSubscribed,
        }, null, 2);
    }
}
exports.SubscriptionState = SubscriptionState;
