// @gomomento/sdk@1.106.0/dist/src/config/transport/storage/transport-strategy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticStorageTransportStrategy = exports.StaticStorageGrpcConfiguration = void 0;
class StaticStorageGrpcConfiguration {
    constructor(props) {
        this.deadlineMillis = props.deadlineMillis;
    }
    getDeadlineMillis() {
        return this.deadlineMillis;
    }
    withDeadlineMillis(deadlineMillis) {
        return new StaticStorageGrpcConfiguration({
            ...this,
            deadlineMillis,
        });
    }
}
exports.StaticStorageGrpcConfiguration = StaticStorageGrpcConfiguration;
class StaticStorageTransportStrategy {
    constructor(props) {
        this.grpcConfig = props.grpcConfiguration;
    }
    getGrpcConfig() {
        return this.grpcConfig;
    }
    withGrpcConfig(grpcConfiguration) {
        return new StaticStorageTransportStrategy({
            ...this,
            grpcConfiguration,
        });
    }
    withClientTimeoutMillis(clientTimeoutMillis) {
        return new StaticStorageTransportStrategy({
            ...this,
            grpcConfiguration: this.grpcConfig.withDeadlineMillis(clientTimeoutMillis),
        });
    }
}
exports.StaticStorageTransportStrategy = StaticStorageTransportStrategy;
