// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/storage/scalar/storage-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.NotFound = exports.Found = void 0;
const enums_1 = require("../../enums");
const response_base_1 = require("../../response-base");
const storage_value_1 = require("./storage-value");
/**
 * Indicates that the item was fetched from the store.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the successful response:
 *
 * - `value()` - the value associated with the key that was retrieved from the cache.
 */
class Found extends response_base_1.ResponseBase {
    /**
     * Creates an instance of the Found response.
     * @param {StorageValue} value - The value associated with the key that was retrieved from the cache.
     */
    constructor(value) {
        super();
        this.type = enums_1.StorageGetResponse.Found;
        this._value = value;
    }
    /**
     * Creates a Found response with an integer value.
     * @param {number} value - The integer value to be stored.
     * @returns {Found} - A Found response object containing the integer value.
     */
    static ofInt(value) {
        return new Found(storage_value_1.StorageValue.ofInt(value));
    }
    /**
     * Creates a Found response with a double value.
     * @param {number} value - The double value to be stored.
     * @returns {Found} - A Found response object containing the double value.
     */
    static ofDouble(value) {
        return new Found(storage_value_1.StorageValue.ofDouble(value));
    }
    /**
     * Creates a Found response with a string value.
     * @param {string} value - The string value to be stored.
     * @returns {Found} - A Found response object containing the string value.
     */
    static ofString(value) {
        return new Found(storage_value_1.StorageValue.ofString(value));
    }
    /**
     * Creates a Found response with a byte array value.
     * @param {Uint8Array} value - The byte array value to be stored.
     * @returns {Found} - A Found response object containing the byte array value.
     */
    static ofBytes(value) {
        return new Found(storage_value_1.StorageValue.ofBytes(value));
    }
    /**
     * Retrieves the value associated with the key that was retrieved from the cache.
     * @returns {StorageValue} - The value associated with the key, or undefined if no value is present.
     */
    value() {
        return this._value;
    }
}
exports.Found = Found;
/**
 * Indicates that the item was not found in the store.
 */
class NotFound extends response_base_1.ResponseBase {
    constructor() {
        super();
        this.type = enums_1.StorageGetResponse.NotFound;
    }
    value() {
        return undefined;
    }
}
exports.NotFound = NotFound;
/**
 * Indicates that an error occurred during the storage get request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends response_base_1.BaseResponseError {
    constructor(_innerException) {
        super(_innerException);
        this.type = enums_1.StorageGetResponse.Error;
    }
    value() {
        return undefined;
    }
}
exports.Error = Error;
