// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/storage/scalar/storage-value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageValue = void 0;
/**
 * Represents a value stored in the cache, which can be an integer, double, string, or byte array.
 */
class StorageValue {
    /**
     * Creates an instance of the StorageValue class.
     * @param {number | undefined} valueInt - The integer value to be stored.
     * @param {number | undefined} valueDouble - The double value to be stored.
     * @param {string | undefined} valueString - The string value to be stored.
     * @param {Uint8Array | undefined} valueBytes - The byte array value to be stored.
     */
    constructor(valueInt, valueDouble, valueString, valueBytes) {
        this._valueInt = undefined;
        this._valueDouble = undefined;
        this._valueString = undefined;
        this._valueBytes = undefined;
        this._valueInt = valueInt;
        this._valueDouble = valueDouble;
        this._valueString = valueString;
        this._valueBytes = valueBytes;
    }
    /**
     * Creates a StorageValue instance with an integer value.
     * @param {number} value - The integer value to be stored.
     * @returns {StorageValue} - A StorageValue instance containing the integer value.
     */
    static ofInt(value) {
        return new StorageValue(value, undefined, undefined, undefined);
    }
    /**
     * Creates a StorageValue instance with a double value.
     * @param {number} value - The double value to be stored.
     * @returns {StorageValue} - A StorageValue instance containing the double value.
     */
    static ofDouble(value) {
        return new StorageValue(undefined, value, undefined, undefined);
    }
    /**
     * Creates a StorageValue instance with a string value.
     * @param {string} value - The string value to be stored.
     * @returns {StorageValue} - A StorageValue instance containing the string value.
     */
    static ofString(value) {
        return new StorageValue(undefined, undefined, value, undefined);
    }
    /**
     * Creates a StorageValue instance with a byte array value.
     * @param {Uint8Array} value - The byte array value to be stored.
     * @returns {StorageValue} - A StorageValue instance containing the byte array value.
     */
    static ofBytes(value) {
        return new StorageValue(undefined, undefined, undefined, value);
    }
    /**
     * Retrieves the integer value stored in this instance.
     * @returns {number | undefined} - The integer value, or undefined if no integer value is present.
     */
    int() {
        return this._valueInt;
    }
    /**
     * Retrieves the double value stored in this instance.
     * @returns {number | undefined} - The double value, or undefined if no double value is present.
     */
    double() {
        return this._valueDouble;
    }
    /**
     * Retrieves the string value stored in this instance.
     * @returns {string | undefined} - The string value, or undefined if no string value is present.
     */
    string() {
        return this._valueString;
    }
    /**
     * Retrieves the byte array value stored in this instance.
     * @returns {Uint8Array | undefined} - The byte array value, or undefined if no byte array value is present.
     */
    bytes() {
        return this._valueBytes;
    }
}
exports.StorageValue = StorageValue;
