// @gomomento/sdk@1.106.0/dist/src/config/storage-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageClientConfiguration = void 0;
class StorageClientConfiguration {
    constructor(props) {
        this.loggerFactory = props.loggerFactory;
        this.transportStrategy = props.transportStrategy;
        this.retryStrategy = props.retryStrategy;
    }
    getLoggerFactory() {
        return this.loggerFactory;
    }
    getTransportStrategy() {
        return this.transportStrategy;
    }
    getRetryStrategy() {
        return this.retryStrategy;
    }
    withClientTimeoutMillis(clientTimeoutMillis) {
        return new StorageClientConfiguration({
            ...this,
            transportStrategy: this.transportStrategy.withClientTimeoutMillis(clientTimeoutMillis),
        });
    }
    withTransportStrategy(transportStrategy) {
        return new StorageClientConfiguration({
            ...this,
            transportStrategy,
        });
    }
    withRetryStrategy(retryStrategy) {
        return new StorageClientConfiguration({
            ...this,
            retryStrategy,
        });
    }
}
exports.StorageClientConfiguration = StorageClientConfiguration;
