// @gomomento/sdk@1.106.0/dist/src/config/storage-configurations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Laptop = void 0;
const default_momento_logger_1 = require("./logging/default-momento-logger");
const storage_configuration_1 = require("./storage-configuration");
const storage_1 = require("./transport/storage");
const fixed_timeout_retry_strategy_1 = require("./retry/fixed-timeout-retry-strategy");
const defaultLoggerFactory = new default_momento_logger_1.DefaultMomentoLoggerFactory();
function defaultRetryStrategy(loggerFactory) {
    return new fixed_timeout_retry_strategy_1.FixedTimeoutRetryStrategy({
        loggerFactory: loggerFactory,
    });
}
/**
 * Laptop config provides defaults suitable for a medium-to-high-latency dev environment.
 * @export
 * @class Laptop
 */
class Laptop extends storage_configuration_1.StorageClientConfiguration {
    /**
     * Provides the latest recommended configuration for a laptop development environment.  NOTE: this configuration may
     * change in future releases to take advantage of improvements we identify for default configurations.
     * @param {MomentoLoggerFactory} [loggerFactory=defaultLoggerFactory]
     * @returns {StorageConfiguration}
     */
    static latest(loggerFactory = defaultLoggerFactory) {
        return new storage_configuration_1.StorageClientConfiguration({
            loggerFactory: loggerFactory,
            transportStrategy: new storage_1.StaticStorageTransportStrategy({
                grpcConfiguration: new storage_1.StaticStorageGrpcConfiguration({
                    deadlineMillis: 5000,
                }),
            }),
            retryStrategy: defaultRetryStrategy(loggerFactory),
        });
    }
}
exports.Laptop = Laptop;
