// @gomomento/sdk@1.106.0/dist/src/config/topic-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopicClientConfiguration = void 0;
class TopicClientConfiguration {
    constructor(props) {
        this.loggerFactory = props.loggerFactory;
        this.transportStrategy = props.transportStrategy;
        this.throwOnErrors = props.throwOnErrors;
    }
    getLoggerFactory() {
        return this.loggerFactory;
    }
    getTransportStrategy() {
        return this.transportStrategy;
    }
    withTransportStrategy(transportStrategy) {
        return new TopicClientConfiguration({
            ...this,
            transportStrategy,
        });
    }
    withNumConnections(numConnections) {
        return this.withTransportStrategy(this.getTransportStrategy().withGrpcConfig(this.getTransportStrategy()
            .getGrpcConfig()
            .withNumClients(numConnections)));
    }
    getThrowOnErrors() {
        return this.throwOnErrors;
    }
    withThrowOnErrors(throwOnErrors) {
        return new TopicClientConfiguration({
            ...this,
            throwOnErrors,
        });
    }
}
exports.TopicClientConfiguration = TopicClientConfiguration;
