// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/enums/cache/scalar/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheDecreaseTtlResponse = exports.CacheIncreaseTtlResponse = exports.CacheUpdateTtlResponse = exports.CacheItemGetTtlResponse = exports.CacheItemGetTypeResponse = exports.CacheKeysExistResponse = exports.CacheKeyExistsResponse = exports.CacheSetIfNotExistsResponse = exports.CacheSetIfPresentAndNotEqualResponse = exports.CacheSetIfAbsentOrEqualResponse = exports.CacheSetIfNotEqualResponse = exports.CacheSetIfEqualResponse = exports.CacheSetIfPresentResponse = exports.CacheSetIfAbsentResponse = exports.CacheSetBatchResponse = exports.CacheGetBatchResponse = exports.CacheIncrementResponse = exports.CacheDeleteResponse = exports.CacheSetResponse = exports.CacheGetResponse = void 0;
var CacheGetResponse;
(function (CacheGetResponse) {
    CacheGetResponse["Hit"] = "Hit";
    CacheGetResponse["Miss"] = "Miss";
    CacheGetResponse["Error"] = "Error";
})(CacheGetResponse = exports.CacheGetResponse || (exports.CacheGetResponse = {}));
var CacheSetResponse;
(function (CacheSetResponse) {
    CacheSetResponse["Success"] = "Success";
    CacheSetResponse["Error"] = "Error";
})(CacheSetResponse = exports.CacheSetResponse || (exports.CacheSetResponse = {}));
var CacheDeleteResponse;
(function (CacheDeleteResponse) {
    CacheDeleteResponse["Success"] = "Success";
    CacheDeleteResponse["Error"] = "Error";
})(CacheDeleteResponse = exports.CacheDeleteResponse || (exports.CacheDeleteResponse = {}));
var CacheIncrementResponse;
(function (CacheIncrementResponse) {
    CacheIncrementResponse["Success"] = "Success";
    CacheIncrementResponse["Error"] = "Error";
})(CacheIncrementResponse = exports.CacheIncrementResponse || (exports.CacheIncrementResponse = {}));
var CacheGetBatchResponse;
(function (CacheGetBatchResponse) {
    CacheGetBatchResponse["Success"] = "Success";
    CacheGetBatchResponse["Error"] = "Error";
})(CacheGetBatchResponse = exports.CacheGetBatchResponse || (exports.CacheGetBatchResponse = {}));
var CacheSetBatchResponse;
(function (CacheSetBatchResponse) {
    CacheSetBatchResponse["Success"] = "Success";
    CacheSetBatchResponse["Error"] = "Error";
})(CacheSetBatchResponse = exports.CacheSetBatchResponse || (exports.CacheSetBatchResponse = {}));
var CacheSetIfAbsentResponse;
(function (CacheSetIfAbsentResponse) {
    CacheSetIfAbsentResponse["Stored"] = "Stored";
    CacheSetIfAbsentResponse["NotStored"] = "NotStored";
    CacheSetIfAbsentResponse["Error"] = "Error";
})(CacheSetIfAbsentResponse = exports.CacheSetIfAbsentResponse || (exports.CacheSetIfAbsentResponse = {}));
var CacheSetIfPresentResponse;
(function (CacheSetIfPresentResponse) {
    CacheSetIfPresentResponse["Stored"] = "Stored";
    CacheSetIfPresentResponse["NotStored"] = "NotStored";
    CacheSetIfPresentResponse["Error"] = "Error";
})(CacheSetIfPresentResponse = exports.CacheSetIfPresentResponse || (exports.CacheSetIfPresentResponse = {}));
var CacheSetIfEqualResponse;
(function (CacheSetIfEqualResponse) {
    CacheSetIfEqualResponse["Stored"] = "Stored";
    CacheSetIfEqualResponse["NotStored"] = "NotStored";
    CacheSetIfEqualResponse["Error"] = "Error";
})(CacheSetIfEqualResponse = exports.CacheSetIfEqualResponse || (exports.CacheSetIfEqualResponse = {}));
var CacheSetIfNotEqualResponse;
(function (CacheSetIfNotEqualResponse) {
    CacheSetIfNotEqualResponse["Stored"] = "Stored";
    CacheSetIfNotEqualResponse["NotStored"] = "NotStored";
    CacheSetIfNotEqualResponse["Error"] = "Error";
})(CacheSetIfNotEqualResponse = exports.CacheSetIfNotEqualResponse || (exports.CacheSetIfNotEqualResponse = {}));
var CacheSetIfAbsentOrEqualResponse;
(function (CacheSetIfAbsentOrEqualResponse) {
    CacheSetIfAbsentOrEqualResponse["Stored"] = "Stored";
    CacheSetIfAbsentOrEqualResponse["NotStored"] = "NotStored";
    CacheSetIfAbsentOrEqualResponse["Error"] = "Error";
})(CacheSetIfAbsentOrEqualResponse = exports.CacheSetIfAbsentOrEqualResponse || (exports.CacheSetIfAbsentOrEqualResponse = {}));
var CacheSetIfPresentAndNotEqualResponse;
(function (CacheSetIfPresentAndNotEqualResponse) {
    CacheSetIfPresentAndNotEqualResponse["Stored"] = "Stored";
    CacheSetIfPresentAndNotEqualResponse["NotStored"] = "NotStored";
    CacheSetIfPresentAndNotEqualResponse["Error"] = "Error";
})(CacheSetIfPresentAndNotEqualResponse = exports.CacheSetIfPresentAndNotEqualResponse || (exports.CacheSetIfPresentAndNotEqualResponse = {}));
var CacheSetIfNotExistsResponse;
(function (CacheSetIfNotExistsResponse) {
    CacheSetIfNotExistsResponse["Stored"] = "Stored";
    CacheSetIfNotExistsResponse["NotStored"] = "NotStored";
    CacheSetIfNotExistsResponse["Error"] = "Error";
})(CacheSetIfNotExistsResponse = exports.CacheSetIfNotExistsResponse || (exports.CacheSetIfNotExistsResponse = {}));
var CacheKeyExistsResponse;
(function (CacheKeyExistsResponse) {
    CacheKeyExistsResponse["Success"] = "Success";
    CacheKeyExistsResponse["Error"] = "Error";
})(CacheKeyExistsResponse = exports.CacheKeyExistsResponse || (exports.CacheKeyExistsResponse = {}));
var CacheKeysExistResponse;
(function (CacheKeysExistResponse) {
    CacheKeysExistResponse["Success"] = "Success";
    CacheKeysExistResponse["Error"] = "Error";
})(CacheKeysExistResponse = exports.CacheKeysExistResponse || (exports.CacheKeysExistResponse = {}));
var CacheItemGetTypeResponse;
(function (CacheItemGetTypeResponse) {
    CacheItemGetTypeResponse["Hit"] = "Hit";
    CacheItemGetTypeResponse["Miss"] = "Miss";
    CacheItemGetTypeResponse["Error"] = "Error";
})(CacheItemGetTypeResponse = exports.CacheItemGetTypeResponse || (exports.CacheItemGetTypeResponse = {}));
var CacheItemGetTtlResponse;
(function (CacheItemGetTtlResponse) {
    CacheItemGetTtlResponse["Hit"] = "Hit";
    CacheItemGetTtlResponse["Miss"] = "Miss";
    CacheItemGetTtlResponse["Error"] = "Error";
})(CacheItemGetTtlResponse = exports.CacheItemGetTtlResponse || (exports.CacheItemGetTtlResponse = {}));
// Note: update ttl has no NotSet response type
// as it sets the ttl unconditionally. We can still
// add this NotSet case later if actually needed.
var CacheUpdateTtlResponse;
(function (CacheUpdateTtlResponse) {
    CacheUpdateTtlResponse["Set"] = "Set";
    CacheUpdateTtlResponse["Miss"] = "Miss";
    CacheUpdateTtlResponse["Error"] = "Error";
})(CacheUpdateTtlResponse = exports.CacheUpdateTtlResponse || (exports.CacheUpdateTtlResponse = {}));
var CacheIncreaseTtlResponse;
(function (CacheIncreaseTtlResponse) {
    CacheIncreaseTtlResponse["Set"] = "Set";
    CacheIncreaseTtlResponse["NotSet"] = "NotSet";
    CacheIncreaseTtlResponse["Miss"] = "Miss";
    CacheIncreaseTtlResponse["Error"] = "Error";
})(CacheIncreaseTtlResponse = exports.CacheIncreaseTtlResponse || (exports.CacheIncreaseTtlResponse = {}));
var CacheDecreaseTtlResponse;
(function (CacheDecreaseTtlResponse) {
    CacheDecreaseTtlResponse["Set"] = "Set";
    CacheDecreaseTtlResponse["NotSet"] = "NotSet";
    CacheDecreaseTtlResponse["Miss"] = "Miss";
    CacheDecreaseTtlResponse["Error"] = "Error";
})(CacheDecreaseTtlResponse = exports.CacheDecreaseTtlResponse || (exports.CacheDecreaseTtlResponse = {}));
