// @gomomento/sdk-core@1.106.0/dist/src/messages/responses/enums/cache/control/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListCachesResponse = exports.FlushCacheResponse = exports.DeleteCacheResponse = exports.CreateCacheResponse = void 0;
var CreateCacheResponse;
(function (CreateCacheResponse) {
    CreateCacheResponse["Success"] = "Success";
    CreateCacheResponse["AlreadyExists"] = "AlreadyExists";
    CreateCacheResponse["Error"] = "Error";
})(CreateCacheResponse = exports.CreateCacheResponse || (exports.CreateCacheResponse = {}));
var DeleteCacheResponse;
(function (DeleteCacheResponse) {
    DeleteCacheResponse["Success"] = "Success";
    DeleteCacheResponse["Error"] = "Error";
})(DeleteCacheResponse = exports.DeleteCacheResponse || (exports.DeleteCacheResponse = {}));
var FlushCacheResponse;
(function (FlushCacheResponse) {
    FlushCacheResponse["Success"] = "Success";
    FlushCacheResponse["Error"] = "Error";
})(FlushCacheResponse = exports.FlushCacheResponse || (exports.FlushCacheResponse = {}));
var ListCachesResponse;
(function (ListCachesResponse) {
    ListCachesResponse["Success"] = "Success";
    ListCachesResponse["Error"] = "Error";
})(ListCachesResponse = exports.ListCachesResponse || (exports.ListCachesResponse = {}));
